File: shared/draughts-engine/src/types/piece.ts
/** The two piece types in international draughts */
export enum PieceType {
  Man = 'man',
  King = 'king',
}

/** The two player colors */
export enum PlayerColor {
  White = 'white',
  Black = 'black',
}

/** A piece on the board */
export interface Piece {
  readonly type: PieceType;
  readonly color: PlayerColor;
}

/** Factory functions for creating pieces */
export const createPiece = (type: PieceType, color: PlayerColor): Piece => ({
  type,
  color,
});

export const createWhiteMan = (): Piece => createPiece(PieceType.Man, PlayerColor.White);
export const createBlackMan = (): Piece => createPiece(PieceType.Man, PlayerColor.Black);
export const createWhiteKing = (): Piece => createPiece(PieceType.King, PlayerColor.White);
export const createBlackKing = (): Piece => createPiece(PieceType.King, PlayerColor.Black);

File: shared/draughts-engine-dart/lib/src/types/piece.dart
/// The two piece types in international draughts.
enum PieceType {
  /// A regular piece (man).
  man,

  /// A promoted piece (king).
  king,
}

/// The two player colors.
enum PlayerColor {
  /// White player (pieces start on squares 1-20).
  white,

  /// Black player (pieces start on squares 31-50).
  black,
}

/// A piece on the board.
class Piece {
  /// Creates a piece with the given [type] and [color].
  const Piece({required this.type, required this.color});

  /// The type of piece (man or king).
  final PieceType type;

  /// The color of the piece.
  final PlayerColor color;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Piece && type == other.type && color == other.color;

  @override
  int get hashCode => Object.hash(type, color);

  @override
  String toString() => 'Piece(${color.name} ${type.name})';
}

/// Factory function for creating pieces.
Piece createPiece(PieceType type, PlayerColor color) =>
    Piece(type: type, color: color);

/// Creates a white man piece.
Piece createWhiteMan() => const Piece(type: PieceType.man, color: PlayerColor.white);

/// Creates a black man piece.
Piece createBlackMan() => const Piece(type: PieceType.man, color: PlayerColor.black);

/// Creates a white king piece.
Piece createWhiteKing() =>
    const Piece(type: PieceType.king, color: PlayerColor.white);

/// Creates a black king piece.
Piece createBlackKing() =>
    const Piece(type: PieceType.king, color: PlayerColor.black);
